layout_2020: Implement automatic minimum size of flex items

stshine opened the pull request:
Implement the algorithm described in
https://drafts.csswg.org/css-flexbox/#min-size-auto.

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

<!-- Either: -->
- [x] There are tests for these changes OR


<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

stshine commented:
@bors-servo try=wpt-2020

bors-servo commented:
:broken_heart: Test failed - [checks-github](https://github.com/servo/servo/actions/runs/5015778754/jobs/8992640258)

stshine commented:
@bors-servo try=wpt-2020

stshine commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 54f9cff65695b82ba4531b3e0b3dd91c1c8c28cd with merge e890584ddf55f1b510134c124b4b47ba751d87d3...

bors-servo commented:
:broken_heart: Test failed - [checks-github](https://github.com/servo/servo/actions/runs/5049092672/jobs/9058794943)

Loirooriol commented on components/layout_2020/style_ext.rs:
```diff
@@ -494,3 +494,30 @@ fn size_to_length(size: &Size) -> LengthPercentageOrAuto {
+    match style.get_position().box_sizing {
+        BoxSizing::ContentBox => (min_block_size / ratio, max_block_size.map(|b| b / ratio)),
+        BoxSizing::BorderBox => (
+            min_block_size / ratio - pbm.padding_border_sums.inline,
```
Is `ratio` for the inner sizes, the outer sizes, or depends on `box-sizing`?

 - If it's for the inner sizes, this case seems wrong since `min_block_size` is an outer size.
 - If it's for the outer sizes, it's the same problem for the `content-box` case
 - If it depends on `box-sizing`, then `transferred_size_suggestion` needs to check the `box-sizing`

stshine commented on components/layout_2020/flexbox/layout.rs:
```diff
@@ -444,7 +445,79 @@ impl<'a> FlexItem<'a> {
+        // https://drafts.csswg.org/css-flexbox/#min-size-auto
+        let automatic_min_size = || {
+            // FIXME(stshine): Consider more situations when auto min size is not needed.
+            if box_style.get_box().overflow_x != Overflow::Visible {
```
Fine.

Loirooriol commented on components/layout_2020/flexbox/layout.rs:
```diff
@@ -444,7 +445,79 @@ impl<'a> FlexItem<'a> {
+                            content_size_suggestion
+                        }
+                    },
+                }
```
I think you are missing this:

> In all cases, the size is clamped by the [maximum](https://drafts.csswg.org/css-sizing-3/#max-width) [main size](https://drafts.csswg.org/css-flexbox-1/#main-size) if it’s [definite](https://drafts.csswg.org/css-flexbox-1/#definite).

stshine replied:
Yeah I already had this fixed :)

stshine commented:
@bors-servo try=wpt-2020

bors-servo commented:
:hourglass: Trying commit 96499094c8bcc19ce2e8de721790d1559fd31829 with merge 49b52d012799e0b6af08beeabbad1cadaf46b9de...

bors-servo commented:
:broken_heart: Test failed - [checks-github](https://github.com/servo/servo/actions/runs/5069198802/jobs/9103432298)

Loirooriol commented on components/layout_2020/flexbox/layout.rs:
```diff
@@ -444,7 +445,81 @@ impl<'a> FlexItem<'a> {
+                            .contents
+                            .inline_size_over_block_size_intrinsic_ratio(box_style)
+                        {
+                            let (min_inline_size, max_inline_size) = transferred_inline_min_max(
```
This uses the computed min/max block sizes, so it needs to check `box-sizing`. Why not simply use `min_size.block` and `max_size.block` like for the transferred size suggestion?

stshine replied:
Because they are for content size, which was truncated by border padding in block direction when `box-sizing: border-box`, thus will give incorrect result.

Loirooriol replied:
I think we precisely want that truncated value.

Loirooriol commented on components/layout_2020/style_ext.rs:
```diff
@@ -494,3 +494,29 @@ fn size_to_length(size: &Size) -> LengthPercentageOrAuto {
+    match style.get_position().box_sizing {
+        BoxSizing::ContentBox => (min_block_size * ratio, max_block_size.map(|b| b * ratio)),
+        BoxSizing::BorderBox => (
+            min_block_size * ratio - pbm.padding_border_sums.inline,
+            max_block_size.map(|b| b * ratio - pbm.padding_border_sums.inline),
```
It the  ratio refers to the inner size, shouldn't this be more like

```rs
(min_block_size - pbm.padding_border_sums.block) * ratio,
max_block_size.map(|b| (b - pbm.padding_border_sums.block) * ratio),
```

stshine replied:
imo max_block_size comes from style so that's what should be transferred, right?

Loirooriol replied:
I'm checking

```html
<!DOCTYPE html>
<div style="display: flex">
  <img src="image" style="box-sizing: border-box; min-height:300px; border: solid 100px">
</div>
<script>
addEventListener("load", () => document.body.append(document.querySelector("img").getBoundingClientRect().width);
</script>
```

where image is 200px wide and 100px tall (ratio of 2).

Your code takes `min-height:300px`, multiplies it by 2, and subtracts the horizontal border. So the inner width is `300*2 - 200 = 400`, and the outer width is 600.

But you can see that in other browsers, the inner width is 200, and the outer width is 400. That's because the inner height is 100, so they preserve the content ratio of 2.

stshine commented:
@bors-servo try=wpt-2020

bors-servo commented:
:hourglass: Trying commit b620961d367aaf797d37fa825a93bbd7e9ece9ad with merge e69aa6f892188f2f516a48038152f02aa678e587...

stshine commented:
yeah I think you are right. the rationale behind the converted min and max size is to keep a reasonable cross size.

bors-servo commented:
:broken_heart: Test failed - [checks-github](https://github.com/servo/servo/actions/runs/5128837331/jobs/9226387023)

Loirooriol reviewed (approved):
Thanks!

stshine committed:
layout_2020: Implement automatic minimum size of flex items

Implement the algorithm described in
https://drafts.csswg.org/css-flexbox/#min-size-auto.

stshine commented:
Commits revised. Thanks!

Loirooriol commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit c56f5ae has been approved by `Loirooriol`

<!-- @bors-servo r=Loirooriol c56f5ae4e7e8e126dc6fb10f3b7ec9e3bec1c4bc -->

bors-servo commented:
:hourglass: Testing commit c56f5ae4e7e8e126dc6fb10f3b7ec9e3bec1c4bc with merge b15bfe6bfe070d3cec404c992e4160ce6cf82889...

bors-servo commented:
:broken_heart: Test failed - [checks-github](https://github.com/servo/servo/actions/runs/5158037463/jobs/9294347295)

stshine committed:
Update test expectations

stshine commented:
Aaah, wrote the wrong file

Loirooriol commented:
@bors-servo retry

CYBAI commented:
might need r+ again because of new commit pushed.

Loirooriol commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 7368673 has been approved by `Loirooriol`

<!-- @bors-servo r=Loirooriol 7368673d2e4d029ff5b1863a97f2eeb8533df446 -->

bors-servo commented:
:hourglass: Testing commit 7368673d2e4d029ff5b1863a97f2eeb8533df446 with merge 0f8824da6d1ec41ac8e779b9238501fddda50e20...

bors-servo commented:
:sunny: Test successful - [checks-github](https://github.com/servo/servo/actions/runs/5162804605/jobs/9301807479)
Approved by: Loirooriol
Pushing 0f8824da6d1ec41ac8e779b9238501fddda50e20 to master...
